### DISCUSSION TITLE
Docs say Next.js runtime is not supported for middleware, but it works

### DISCUSSION THREAD

**Discussion by yandodov** [vercel/next.js#69503] [Help] [answered]
### Summary

The Next.js docs [say](https://nextjs.org/docs/app/building-your-application/routing/middleware#runtime):

> Middleware currently only supports the [Edge runtime](https://nextjs.org/docs/app/building-your-application/rendering/edge-and-nodejs-runtimes). The Node.js runtime can not be used.

…however, I can add the following `middleware.ts`:

```ts
import type { NextRequest } from "next/server";
import { NextResponse } from "next/server";

export function middleware(request: NextRequest) {
	if (request.nextUrl.pathname.startsWith("/foo")) {
		return NextResponse.redirect(new URL("/bar", request.url));
	}
}
```

…and when I run `npm run dev` and open `http://localhost:3000/foo` it redirects me as expected.

Am I missing something or the docs are misleading?

### Additional information

```js
Operating System:
  Platform: darwin
  Arch: x64
  Version: Darwin Kernel Version 23.6.0: Mon Jul 29 21:13:00 PDT 2024; root:xnu-10063.141.2~1/RELEASE_X86_64
  Available memory (MB): 32768
  Available CPU cores: 16
Binaries:
  Node: 20.16.0
  npm: 10.8.1
  Yarn: 1.22.19
  pnpm: 9.8.0
Relevant Packages:
  next: 14.2.5
  eslint-config-next: 14.2.5
  react: 18.3.1
  react-dom: 18.3.1
  typescript: 5.5.4
Next.js Config:
  output: N/A
```

**Comment by Juneezee** [comment, marked as answer]
Your code works because you are not using any Node.js APIs that are not available in the Edge Runtime. See the **Unsupported APIs** section of Edge Runtime [^1].

[^1]: https://nextjs.org/docs/app/api-reference/edge#unsupported-apis

**Reply by yandodov** [reply]
But the docs don't mention anything about support, nor APIs…

When I read this:

> The Node.js runtime can not be used.

…I understand that middleware can not be used when running Next.js in Node.

This would be infinitely clearer:

> Middleware currently only supports APIs compatible with the [Edge runtime](https://nextjs.org/docs/app/building-your-application/rendering/edge-and-nodejs-runtimes). APIs exclusive to Node.js [are unsupported](https://nextjs.org/docs/app/api-reference/edge#unsupported-apis).

**Comment by leerob** [comment]
Good news everyone, **Node.js Middleware support landed in Next.js 15.2** (experimental). This is now ready for testing. For any issues you run into, please open new issues so we can track and address before marking it as stable. Thank you!

https://nextjs.org/blog/next-15-2#nodejs-middleware-experimental

**Reply by yandodov** [reply]
As in the PR itself (see above), I'm confused here as well. We're *already using* middleware in Node.js. Does this feature mean that Node.js *APIs* will be supported *on the edge*? In my mind, that's a different thing.

**Reply by leerob** [reply]
Edge was previously the only support runtime for Middleware. You could use a subset of Node.js APIs, but not all of them. With this release, you can now use any Node.js APIs or community packages in Middleware. This runtime support is different from how the application infrastructure is deployed (e.g. "on the edge").

When you deploy a self-hosted Next.js app, you are doing to an origin server. If you are asking about Vercel, for example, we will run the Node.js functions in every region for Middleware, correct.
